util/list_test.go

```diff
@@ -0,0 +1,228 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package util
+
+import (
+	"testing"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+)
+
+func checkLen[T any](list *List[T]) int {
+	if list.root.next == &list.root {
+		return 0
+	}
+
+	return countListSize(&list.root)
+}
+
+func countListSize[T any](head *ListNode[T]) (counter int) {
+	for i := head.next; i != head && i != nil; i = i.next {
+		counter++
+	}
+	return counter
+}
+
+func checkListLen[T any](t *testing.T, l *List[T], len int) bool {
+	if n := checkLen(l); n != len {
+		t.Errorf("l.Len() = %d, want %d", n, len)
+		return true
+	}
+	return false
+}
+
+func checkListPointers[T any](t *testing.T, l *List[T], es []*ListNode[T]) {
+	root := &l.root
+
+	if failed := checkListLen(t, l, len(es)); failed {
+		return
+	}
+
+	if failed := zeroListInspection(t, l, len(es), root); failed {
+		return
+	}
+
+	pointerInspection(t, es, root)
+}
+
+func zeroListInspection[T any](t *testing.T, l *List[T], len int, root *ListNode[T]) bool {
+	// zero length lists must be the zero value or properly initialized (sentinel circle)
+	if len == 0 {
+		if l.root.next != nil && l.root.next != root || l.root.prev != nil && l.root.prev != root {
+			t.Errorf("l.root.next = %p, l.root.prev = %p; both should both be nil or %p", l.root.next, l.root.prev, root)
+		}
+		return true
+	}
+	return false
+}
+
+func pointerInspection[T any](t *testing.T, es []*ListNode[T], root *ListNode[T]) {
+	// check internal and external prev/next connections
+	for i, e := range es {
+		prev := root
+		if i > 0 {
+			prev = es[i-1]
+		}
+		if p := e.prev; p != prev {
+			t.Errorf("elt[%d](%p).prev = %p, want %p", i, e, p, prev)
+		}
+
+		next := root
+		if i < len(es)-1 {
+			next = es[i+1]
+		}
+		if n := e.next; n != next {
+			t.Errorf("elt[%d](%p).next = %p, want %p", i, e, n, next)
+		}
+	}
+}
+
+type testVal struct {
+	val int
+}
+
+func TestList_RemoveFromList(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	l := NewList[testVal]()
+	e1 := l.PushFront(&testVal{1})
+	e2 := l.PushFront(&testVal{2})
+	e3 := l.PushFront(&testVal{3})
+	checkListPointers(t, l, []*ListNode[testVal]{e3, e2, e1})
+
+	l.Remove(e2)
+	checkListPointers(t, l, []*ListNode[testVal]{e3, e1})
+	l.Remove(e3)
+	checkListPointers(t, l, []*ListNode[testVal]{e1})
+}
+
+func TestList_AddingNewNodeWithAllocatedFreeList(t *testing.T) {
```

Comment:
Maybe this came from a previous test file, but there shouldn't be an underscore, right?

Reply:
I kind of prefer underscores because of noticeable prefix, so it is easy to run only subset of tests with `-run TestList_`